src/tactic/equiv_rw.lean

```diff
@@ -0,0 +1,287 @@
+
+-- TODO: We should also use `category_theory.functorial` and `category_theory.hygienic` instances.
+-- (example goal: we could rewrite along an isomorphism of rings (either as `R ≅ S` or `R ≃+* S`)
+-- and turn an `x : mv_polynomial σ R` into an `x : mv_polynomial σ S`.).
+
+meta def equiv_congr_lemmas : tactic (list expr) :=
+do exprs ←
+  [
+  `equiv.of_iff,
+  -- These are functorial w.r.t equiv,
+  -- and so will be subsumed by `equiv_functor.map_equiv` in a subsequent PR.
+  -- Many more could be added, but will wait for that framework.
+  `equiv.perm_congr, `equiv.equiv_congr, `equiv.unique_congr,
+  -- The function arrow is technically a bifunctor `Typeᵒᵖ → Type → Type`,
+  -- but the pattern matcher will never see this.
+  `equiv.arrow_congr',
+  -- Allow rewriting in subtypes:
+  `equiv.subtype_equiv_of_subtype',
+  -- Allow rewriting in the first component of a sigma-type:
+  `equiv.sigma_congr_left',
+  -- Allow rewriting ∀s:
+  -- (You might think that repeated application of `equiv.forall_congr'
+  -- would handle the higher arity cases, but unfortunately unification is not clever enough.)
+  `equiv.forall₃_congr',
+  `equiv.forall₂_congr',
+  `equiv.forall_congr',
+  -- Allow rewriting in argument of Pi types:
+   `equiv.Pi_congr_left',
+  -- Handles `sum` and `prod`, and many others:
+  `bifunctor.map_equiv,
+  -- Handles `list`, `option`, and many others:
+  `functor.map_equiv,
+  -- We have to filter results to ensure we don't cheat and use exclusively `equiv.refl` and `iff.refl`!
+  `equiv.refl,
+  `iff.refl
+  ].mmap (λ n, try_core (mk_const n)),
+  return (exprs.map option.to_list).join -- TODO: implement `.mfilter_map mk_const`?
+
+declare_trace equiv_rw_type
+
+/--
+Configuration structure for `equiv_rw`.
+
+* `max_steps` bounds the search depth for equivalences to rewrite along.
+  The default value is 10.
+  (e.g., if you're rewriting along `e : α ≃ β`, and `max_steps := 2`,
+  you can rewrite `option (option α))` but not `option (option (option α))`.
+-/
+meta structure equiv_rw_cfg :=
+(max_steps : ℕ := 10)
+
+/--
+Implementation of `equiv_rw_type`, using `solve_by_elim`.
+Expects a goal of the form `t ≃ _`,
+and tries to solve it using `eq : α ≃ β` and congruence lemmas.
+-/
+meta def equiv_rw_type_core (eq : expr) (cfg : equiv_rw_cfg) : tactic unit :=
+do
+  -- Assemble the relevant lemmas.
+  equiv_congr_lemmas ← equiv_congr_lemmas,
+  /-
+    We now call `solve_by_elim` to try to generate the requested equivalence.
+    There are a few subtleties!
+    * We make sure that `eq` is the first lemma, so it is applied whenever possible.
+    * In `equiv_congr_lemmas`, we put `equiv.refl` last so it is only used when it is not possible
+      to descend further.
+    * Since some congruence lemmas generate subgoals with `∀` statements,
+      we use the `pre_apply` subtactic of `solve_by_elim` to preprocess each new goal with `intros`.
```

Comment:
It seems to me like `solve_by_elim` should do this by default.

Reply:
This might be dangerous. @jcommelin recently tried using `solve_by_elim` in a way that would have been broken by this. He had a goal `a ∧ b → c ∧ d` and was hitting it with `solve_by_elim [and.imp, ...]`. If we'd called `intros` automatically, the goal would have just become `c ∧ d`, and when we applied `and.imp` it would have given us metavariables for `a` and `b`, and then not known what to do.

Perhaps we should discourage that use in any case, as I agree `intros` along the way is tempting.

For now I'm not going to make this change to `solve_by_elim` in this PR, but I'm happy to do it separately.

Reply:
I think it's hard to make heuristics for when you should avoid `intros` and use `and.imp` and `or.imp` instead. So maybe we can (in another PR) extends `solve_by_elim`s powers to include `intros`, and have a `solve_by_elim'` that could still be used for special cases like mine.